bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java

```diff
@@ -0,0 +1,252 @@
+    private static final String CONFIG_MAX_THREADS_CUSTOM = "maxThreadsCustom";
+    private static final String CONFIG_KEEP_ALIVE_CUSTOM = "keepAliveTimeoutCustom";
+
+    private static final int MIN_CONSUMER_NAME_LENGTH = 4;
+    private static final int MAX_CONSUMER_NAME_LENGTH = 20;
+    private static final Pattern CONSUMER_NAME_PATTERN = Pattern.compile("[a-zA-Z0-9_\\-]*");
+
+    private volatile TrustManagerProvider trustmanagerProvider;
+
+    private QueuedThreadPool threadPool = null;
+    private HttpClient sharedHttpClient = null;
+    private CommonHttpClient sharedHttpClientFacade = null;
+
+    private int minThreadsShared;
+    private int maxThreadsShared;
+    private int keepAliveTimeoutShared;
+    private int minThreadsCustom;
+    private int maxThreadsCustom;
+    private int keepAliveTimeoutCustom;
+
+    @Activate
+    protected void activate(Map<String, Object> parameters) {
+        getConfigParameters(parameters);
+    }
+
+    @Modified
+    protected void modified(Map<String, Object> parameters) {
+        getConfigParameters(parameters);
+        if (threadPool != null) {
+            threadPool.setMinThreads(minThreadsShared);
+            threadPool.setMaxThreads(maxThreadsShared);
+            threadPool.setIdleTimeout(keepAliveTimeoutShared * 1000);
+        }
+    }
+
+    @Deactivate
+    protected void deactivate() {
+        try {
+            if (sharedHttpClient != null) {
+                sharedHttpClient.stop();
+                sharedHttpClientFacade = null;
+                sharedHttpClient = null;
+                threadPool = null;
+                logger.info("jetty shared http client stopped");
+            }
+        } catch (Exception e) {
+            logger.error("error while stppping jetty shared http client", e);
+            // nothing else we can do here
+        }
+
+    }
+
+    @Override
+    public HttpClient createHttpClient(String consumerName, String endpoint) {
+        Objects.requireNonNull(endpoint, "endpoint must not be null");
+        logger.debug("httpClient for endpoint {} requested", endpoint);
+        checkConsumerName(consumerName);
+        return createHttpClientInternal(consumerName, endpoint);
+    }
+
+    @Override
+    public CommonHttpClient getHttpClient() {
+        initialize();
+        logger.debug("shared httpClient requested");
+        return sharedHttpClientFacade;
+    }
+
+    private void getConfigParameters(Map<String, Object> parameters) {
+        minThreadsShared = (int) parameters.get(CONFIG_MIN_THREADS_SHARED);
+        maxThreadsShared = (int) parameters.get(CONFIG_MAX_THREADS_SHARED);
+        keepAliveTimeoutShared = (int) parameters.get(CONFIG_KEEP_ALIVE_SHARED);
+        minThreadsCustom = (int) parameters.get(CONFIG_MIN_THREADS_CUSTOM);
+        maxThreadsCustom = (int) parameters.get(CONFIG_MAX_THREADS_CUSTOM);
+        keepAliveTimeoutCustom = (int) parameters.get(CONFIG_KEEP_ALIVE_CUSTOM);
+    }
+
+    private void initialize() {
```

Comment:
I would think that you should have some kind of synchronization/locking to avoid that several threads are requesting the client concurrently and executing the initialize method in parallel.

Reply:
The run method of the doPrivileged call is synchonized

Reply:
Err, what does that help? Every thread has its own instance of `PrivilegedExceptionAction`, so the `synchronized` won't have any effect here or do I miss something?

Reply:
But I changed it a little bit, so that the doPrivileged call is only done if needed.

Reply:
Already changed